nautobot/core/jobs/__init__.py

```diff
@@ -361,6 +371,131 @@ def run(self, *, content_type, csv_data=None, csv_file=None, roll_back_if_error=
             raise RunJobTaskFailed("CSV import not fully successful, see logs")
 
 
+def get_data_compliance_rules():
+    """Generate a list of Audit Ruleset classes that exist from the registry."""
+    validators = []
+    for rule_sets in get_data_compliance_rules_map().values():
+        validators.extend(rule_sets)
+    return validators
+
+
+def get_choices():
+    """Get choices from registry."""
+    choices = []
+    for ruleset_class in get_data_compliance_rules():
+        choices.append((ruleset_class.__name__, ruleset_class.__name__))
+    for repo in GitRepository.objects.all():
+        if "nautobot_data_validation_engine.data_compliance_rules" in repo.provided_contents:
+            for compliance_class in get_classes_from_git_repo(repo):
+                choices.append((compliance_class.__name__, compliance_class.__name__))
+
+    choices.sort()
+    return choices
+
+
+def clean_compliance_rules_results_for_instance(instance, excluded_pks):
+    """Clean compliance results."""
+    excluded_pks = excluded_pks or []
+    DataCompliance.objects.filter(
+        object_id=instance.id,
+        content_type=ContentType.objects.get_for_model(instance),
+        compliance_class_name__endswith="CustomValidator",
+    ).exclude(pk__in=excluded_pks).delete()
+
+
+class RunRegisteredDataComplianceRules(Job):
+    """Run the validate function on all registered DataComplianceRule classes and, optionally, the built-in data validation rules."""
+
+    name = "Run Registered Data Compliance Rules"
+    description = "Runs selected Data Compliance rule classes."
+
+    selected_data_compliance_rules = MultiChoiceVar(
+        choices=get_choices,
+        label="Select Data Compliance Rules",
+        required=False,
+        description="Not selecting any rules will run all rules listed.",
+    )
+
+    run_builtin_rules_in_report = BooleanVar(
+        label="Run built-in validation rules?", description="Include created, built-in data validation rules in report."
+    )
+
+    def run(self, *args, **kwargs):
+        """Run the validate function on all given DataComplianceRule classes."""
+        selected_data_compliance_rules = kwargs.get("selected_data_compliance_rules", None)
+
+        compliance_classes = []
+        compliance_classes.extend(get_data_compliance_rules())
+
+        for repo in GitRepository.objects.all():
+            if "nautobot_data_validation_engine.data_compliance_rules" in repo.provided_contents:
+                compliance_classes.extend(get_classes_from_git_repo(repo))
+
+        for compliance_class in compliance_classes:
+            if selected_data_compliance_rules and compliance_class.__name__ not in selected_data_compliance_rules:
+                continue
+            self.logger.info(f"Running {compliance_class.__name__}")
+            app_label, model = compliance_class.model.split(".")
+            for obj in global_apps.get_model(app_label, model).objects.all():
```

Comment:
Possibly use `.iterator()` here to reduce overhead when there's a lot of objects?